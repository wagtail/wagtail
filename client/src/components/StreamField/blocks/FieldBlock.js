/* global $ */

import { escapeHtml as h } from '../../../utils/text';

export class FieldBlock {
  constructor(blockDef, placeholder, prefix, initialState, initialError) {
    this.blockDef = blockDef;
    this.type = blockDef.name;

    const dom = $(`
      <div class="${h(this.blockDef.meta.classname)}">
        <div class="field-content">
          <div class="input">
            <div data-streamfield-widget></div>
            <span></span>
          </div>
        </div>
      </div>
    `);
    $(placeholder).replaceWith(dom);
    const widgetElement = dom.find('[data-streamfield-widget]').get(0);
    this.element = dom[0];

    try {
      this.widget = this.blockDef.widget.render(
        widgetElement,
        prefix,
        prefix,
        initialState,
      );
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e);
      this.setError([
        {
          messages: [
            'This widget failed to render, please check the console for details',
          ],
        },
      ]);
      return;
    }

    this.idForLabel = this.widget.idForLabel;

    if (this.blockDef.meta.helpText) {
      const helpElement = document.createElement('p');
      helpElement.classList.add('help');
      helpElement.innerHTML = this.blockDef.meta.helpText; // unescaped, as per Django conventions
      this.element.querySelector('.field-content').appendChild(helpElement);
    }

    if (window.comments && this.blockDef.meta.showAddCommentButton) {
      const fieldCommentControlElement = document.createElement('div');
      fieldCommentControlElement.classList.add('field-comment-control');
      this.element.appendChild(fieldCommentControlElement);

      const addCommentButtonElement = document.createElement('button');
      addCommentButtonElement.type = 'button';
      addCommentButtonElement.setAttribute(
        'aria-label',
        blockDef.meta.strings.ADD_COMMENT,
      );
      addCommentButtonElement.setAttribute('data-comment-add', '');
      addCommentButtonElement.classList.add('button');
      addCommentButtonElement.classList.add('button-secondary');
      addCommentButtonElement.classList.add('button-small');
      addCommentButtonElement.classList.add('u-hidden');
      addCommentButtonElement.innerHTML =
        '<svg class="icon icon-comment-add initial icon-default" aria-hidden="true" focusable="false">' +
        '<use href="#icon-comment-add"></use></svg>' +
        '<svg class="icon icon-comment-add initial icon-reversed" aria-hidden="true" focusable="false">' +
        '<use href="#icon-comment-add-reversed"></use></svg>';
      fieldCommentControlElement.appendChild(addCommentButtonElement);
      window.comments.initAddCommentButton(addCommentButtonElement);
    }

    if (initialError) {
      this.setError(initialError);
    }
  }

  setState(state) {
    if (this.widget) {
      this.widget.setState(state);
    }
  }

  setError(errorList) {
    this.element
      .querySelectorAll(':scope > .field-content > .error-message')
      .forEach((element) => element.remove());

    if (errorList) {
      this.element.classList.add('error');

      const errorElement = document.createElement('p');
      errorElement.classList.add('error-message');
      errorElement.innerHTML = errorList
        .map((error) => `<span>${h(error.messages[0])}</span>`)
        .join('');
      this.element.querySelector('.field-content').appendChild(errorElement);
    } else {
      this.element.classList.remove('error');
    }
  }

  getState() {
    return this.widget.getState();
  }

  getValue() {
    return this.widget.getValue();
  }

  getTextLabel(opts) {
    if (this.widget.getTextLabel) {
      return this.widget.getTextLabel(opts);
    }
    return null;
  }

  focus(opts) {
    if (this.widget) {
      this.widget.focus(opts);
    }
  }
}

export class FieldBlockDefinition {
  constructor(name, widget, meta) {
    this.name = name;
    this.widget = widget;
    this.meta = meta;
  }

  render(placeholder, prefix, initialState, initialError) {
    return new FieldBlock(
      this,
      placeholder,
      prefix,
      initialState,
      initialError,
    );
  }
}
