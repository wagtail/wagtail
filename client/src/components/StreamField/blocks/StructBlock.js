/* global $ */

import { escapeHtml as h } from '../../../utils/text';

export class StructBlockValidationError {
  constructor(blockErrors) {
    this.blockErrors = blockErrors;
  }
}

export class StructBlock {
  constructor(blockDef, placeholder, prefix, initialState, initialError) {
    const state = initialState || {};
    this.blockDef = blockDef;
    this.type = blockDef.name;

    this.childBlocks = {};

    if (blockDef.meta.formTemplate) {
      const html = blockDef.meta.formTemplate.replace(/__PREFIX__/g, prefix);
      const dom = $(html);
      $(placeholder).replaceWith(dom);
      this.blockDef.childBlockDefs.forEach((childBlockDef) => {
        const childBlockElement = dom
          .find('[data-structblock-child="' + childBlockDef.name + '"]')
          .get(0);
        const childBlock = childBlockDef.render(
          childBlockElement,
          prefix + '-' + childBlockDef.name,
          state[childBlockDef.name],
          initialError?.blockErrors[childBlockDef.name],
        );
        this.childBlocks[childBlockDef.name] = childBlock;
      });
    } else {
      const dom = $(`
        <div class="${h(this.blockDef.meta.classname || '')}">
        </div>
      `);
      $(placeholder).replaceWith(dom);

      if (this.blockDef.meta.helpText) {
        // help text is left unescaped as per Django conventions
        dom.append(`
          <span>
            <div class="help">
              ${this.blockDef.meta.helpIcon}
              ${this.blockDef.meta.helpText}
            </div>
          </span>
        `);
      }

      this.blockDef.childBlockDefs.forEach((childBlockDef) => {
        const childDom = $(`
          <div class="field ${
            childBlockDef.meta.required ? 'required' : ''
          }" data-contentpath="${childBlockDef.name}">
            <label class="field__label">${h(childBlockDef.meta.label)}</label>
            <div data-streamfield-block></div>
          </div>
        `);
        dom.append(childDom);
        const childBlockElement = childDom
          .find('[data-streamfield-block]')
          .get(0);
        const labelElement = childDom.find('label').get(0);
        const childBlock = childBlockDef.render(
          childBlockElement,
          prefix + '-' + childBlockDef.name,
          state[childBlockDef.name],
          initialError?.blockErrors[childBlockDef.name],
        );

        this.childBlocks[childBlockDef.name] = childBlock;
        if (childBlock.idForLabel) {
          labelElement.setAttribute('for', childBlock.idForLabel);
        }
      });
    }
  }

  setState(state) {
    // eslint-disable-next-line guard-for-in
    for (const name in state) {
      this.childBlocks[name].setState(state[name]);
    }
  }

  setError(errorList) {
    if (errorList.length !== 1) {
      return;
    }
    const error = errorList[0];

    for (const blockName in error.blockErrors) {
      if (error.blockErrors.hasOwnProperty(blockName)) {
        this.childBlocks[blockName].setError(error.blockErrors[blockName]);
      }
    }
  }

  getState() {
    const state = {};
    // eslint-disable-next-line guard-for-in
    for (const name in this.childBlocks) {
      state[name] = this.childBlocks[name].getState();
    }
    return state;
  }

  getValue() {
    const value = {};
    // eslint-disable-next-line guard-for-in
    for (const name in this.childBlocks) {
      value[name] = this.childBlocks[name].getValue();
    }
    return value;
  }

  getTextLabel(opts) {
    if (this.blockDef.meta.labelFormat) {
      /* use labelFormat - regexp replace any field references like '{first_name}'
      with the text label of that sub-block */
      return this.blockDef.meta.labelFormat.replace(
        /\{(\w+)\}/g,
        (tag, blockName) => {
          const block = this.childBlocks[blockName];
          if (block.getTextLabel) {
            /* to be strictly correct, we should be adjusting opts.maxLength to account for the overheads
          in the format string, and dividing the remainder across all the placeholders in the string,
          rather than just passing opts on to the child. But that would get complicated, and this is
          better than nothing... */
            return block.getTextLabel(opts);
          }
          return '';
        },
      );
    }

    /* if no labelFormat specified, just try each child block in turn until we find one that provides a label */
    for (const childDef of this.blockDef.childBlockDefs) {
      const child = this.childBlocks[childDef.name];
      if (child.getTextLabel) {
        const val = child.getTextLabel(opts);
        if (val) return val;
      }
    }
    // no usable label found
    return null;
  }

  focus(opts) {
    if (this.blockDef.childBlockDefs.length) {
      const firstChildName = this.blockDef.childBlockDefs[0].name;
      this.childBlocks[firstChildName].focus(opts);
    }
  }
}

export class StructBlockDefinition {
  constructor(name, childBlockDefs, meta) {
    this.name = name;
    this.childBlockDefs = childBlockDefs;
    this.meta = meta;
  }

  render(placeholder, prefix, initialState, initialError) {
    return new StructBlock(
      this,
      placeholder,
      prefix,
      initialState,
      initialError,
    );
  }
}
