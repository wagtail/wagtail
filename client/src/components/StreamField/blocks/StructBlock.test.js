import $ from 'jquery';
import { FieldBlockDefinition } from './FieldBlock';
import { StreamBlockDefinition } from './StreamBlock';
import { StructBlockDefinition } from './StructBlock';

window.$ = $;

window.comments = {
  getContentPath: jest.fn(),
};

// Define some callbacks in global scope that can be mocked in tests
let constructor = (_widgetName, _name, _id, _initialState) => {};
let setState = (_widgetName, _state) => {};
let getState = (_widgetName) => {};
let getValue = (_widgetName) => {};
let focus = (_widgetName) => {};

class DummyWidgetDefinition {
  constructor(widgetName) {
    this.widgetName = widgetName;
  }

  render(placeholder, name, id, initialState) {
    const widgetName = this.widgetName;
    constructor(widgetName, { name, id, initialState });

    $(placeholder).replaceWith(
      `<p name="${name}" id="${id}">${widgetName}</p>`,
    );
    return {
      setState(state) {
        setState(widgetName, state);
      },
      getState() {
        getState(widgetName);
        return `state: ${widgetName} - ${name}`;
      },
      getValue() {
        getValue(widgetName);
        return `value: ${widgetName} - ${name}`;
      },
      getTextLabel() {
        return `label: ${name}`;
      },
      focus() {
        focus(widgetName);
      },
      idForLabel: id,
    };
  }
}

describe('telepath: wagtail.blocks.StructBlock', () => {
  let boundBlock;

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Define a test block
    const blockDef = new StructBlockDefinition(
      'heading_block',
      [
        new FieldBlockDefinition(
          'heading_text',
          new DummyWidgetDefinition('Heading widget'),
          {
            label: 'Heading text',
            required: true,
            icon: 'placeholder',
            classname: 'w-field w-field--char_field w-field--text_input',
          },
        ),
        new FieldBlockDefinition(
          'size',
          new DummyWidgetDefinition('Size widget'),
          {
            label: 'Size',
            required: false,
            icon: 'placeholder',
            classname: 'w-field w-field--choice_field w-field--select',
          },
        ),
      ],
      {
        label: 'Heading block',
        required: false,
        icon: 'title',
        classname: 'struct-block',
        helpText: 'use <strong>lots</strong> of these',
        helpIcon: '<div class="icon-help">?</div>',
      },
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render($('#placeholder'), 'the-prefix', {
      heading_text: 'Test heading text',
      size: '123',
    });
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('Widget constructors are called with correct parameters', () => {
    expect(constructor.mock.calls.length).toBe(2);

    expect(constructor.mock.calls[0][0]).toBe('Heading widget');
    expect(constructor.mock.calls[0][1]).toEqual({
      name: 'the-prefix-heading_text',
      id: 'the-prefix-heading_text',
      initialState: 'Test heading text',
    });

    expect(constructor.mock.calls[1][0]).toBe('Size widget');
    expect(constructor.mock.calls[1][1]).toEqual({
      name: 'the-prefix-size',
      id: 'the-prefix-size',
      initialState: '123',
    });
  });

  test('getValue() calls getValue() on all widgets', () => {
    const value = boundBlock.getValue();
    expect(getValue.mock.calls.length).toBe(2);
    expect(value).toEqual({
      heading_text: 'value: Heading widget - the-prefix-heading_text',
      size: 'value: Size widget - the-prefix-size',
    });
  });

  test('getState() calls getState() on all widgets', () => {
    const state = boundBlock.getState();
    expect(getState.mock.calls.length).toBe(2);
    expect(state).toEqual({
      heading_text: 'state: Heading widget - the-prefix-heading_text',
      size: 'state: Size widget - the-prefix-size',
    });
  });

  test('setState() calls setState() on all widgets', () => {
    boundBlock.setState({
      heading_text: 'Changed heading text',
      size: '456',
    });
    expect(setState.mock.calls.length).toBe(2);
    expect(setState.mock.calls[0][0]).toBe('Heading widget');
    expect(setState.mock.calls[0][1]).toBe('Changed heading text');
    expect(setState.mock.calls[1][0]).toBe('Size widget');
    expect(setState.mock.calls[1][1]).toBe('456');
  });

  test('focus() calls focus() on first widget', () => {
    boundBlock.focus();
    expect(focus.mock.calls.length).toBe(1);
    expect(focus.mock.calls[0][0]).toBe('Heading widget');
  });

  test('getTextLabel() returns text label of first widget', () => {
    expect(boundBlock.getTextLabel()).toBe('label: the-prefix-heading_text');
  });

  test('setError passes error messages to children', () => {
    boundBlock.setError({
      blockErrors: {
        size: { messages: ['This is too big'] },
      },
    });
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('setError shows non-block errors', () => {
    boundBlock.setError({
      messages: ['This is just generally wrong'],
    });
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('telepath: wagtail.blocks.StructBlock with formTemplate', () => {
  let boundBlock;
  let blockDefWithBadLabelFormat;

  beforeEach(() => {
    // Create mocks for callbacks
    constructor = jest.fn();
    setState = jest.fn();
    getState = jest.fn();
    getValue = jest.fn();
    focus = jest.fn();

    // Define a test block
    const blockOpts = {
      label: 'Heading block',
      required: false,
      icon: 'title',
      formTemplate: `<div class="custom-form-template">
        <p>here comes the first field:</p>
        <div data-structblock-child="heading_text"></div>
        <p>and here is the second:</p>
        <div data-structblock-child="size"></div>
      </div>`,
      labelFormat: '{heading_text} - {size}',
    };
    const headingTextBlockDef = new FieldBlockDefinition(
      'heading_text',
      new DummyWidgetDefinition('Heading widget'),
      {
        label: 'Heading text',
        required: true,
        icon: 'placeholder',
        classname: 'w-field w-field--char_field w-field--text_input',
      },
    );
    const sizeBlockDef = new FieldBlockDefinition(
      'size',
      new DummyWidgetDefinition('Size widget'),
      {
        label: 'Size',
        required: false,
        icon: 'placeholder',
        classname: 'w-field w-field--choice_field w-field--select',
      },
    );

    const blockDef = new StructBlockDefinition(
      'heading_block',
      [headingTextBlockDef, sizeBlockDef],
      blockOpts,
    );
    blockDefWithBadLabelFormat = new StructBlockDefinition(
      'heading_block',
      [headingTextBlockDef, sizeBlockDef],
      { ...blockOpts, labelFormat: '{bad_variable} - {size}' },
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render($('#placeholder'), 'the-prefix', {
      heading_text: 'Test heading text',
      size: '123',
    });
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('Widget constructors are called with correct parameters', () => {
    expect(constructor.mock.calls.length).toBe(2);

    expect(constructor.mock.calls[0][0]).toBe('Heading widget');
    expect(constructor.mock.calls[0][1]).toEqual({
      name: 'the-prefix-heading_text',
      id: 'the-prefix-heading_text',
      initialState: 'Test heading text',
    });

    expect(constructor.mock.calls[1][0]).toBe('Size widget');
    expect(constructor.mock.calls[1][1]).toEqual({
      name: 'the-prefix-size',
      id: 'the-prefix-size',
      initialState: '123',
    });
  });

  test('getValue() calls getValue() on all widgets', () => {
    const value = boundBlock.getValue();
    expect(getValue.mock.calls.length).toBe(2);
    expect(value).toEqual({
      heading_text: 'value: Heading widget - the-prefix-heading_text',
      size: 'value: Size widget - the-prefix-size',
    });
  });

  test('getState() calls getState() on all widgets', () => {
    const state = boundBlock.getState();
    expect(getState.mock.calls.length).toBe(2);
    expect(state).toEqual({
      heading_text: 'state: Heading widget - the-prefix-heading_text',
      size: 'state: Size widget - the-prefix-size',
    });
  });

  test('setState() calls setState() on all widgets', () => {
    boundBlock.setState({
      heading_text: 'Changed heading text',
      size: '456',
    });
    expect(setState.mock.calls.length).toBe(2);
    expect(setState.mock.calls[0][0]).toBe('Heading widget');
    expect(setState.mock.calls[0][1]).toBe('Changed heading text');
    expect(setState.mock.calls[1][0]).toBe('Size widget');
    expect(setState.mock.calls[1][1]).toBe('456');
  });

  test('focus() calls focus() on first widget', () => {
    boundBlock.focus();
    expect(focus.mock.calls.length).toBe(1);
    expect(focus.mock.calls[0][0]).toBe('Heading widget');
  });

  test('getTextLabel() returns text label according to labelFormat', () => {
    expect(boundBlock.getTextLabel()).toBe(
      'label: the-prefix-heading_text - label: the-prefix-size',
    );
  });

  test('getTextLabel() gracefully handles bad variables in labelFormat', () => {
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDefWithBadLabelFormat.render(
      $('#placeholder'),
      'the-prefix',
      {
        heading_text: 'Test heading text',
        size: '123',
      },
    );
    expect(boundBlock.getTextLabel()).toBe(' - label: the-prefix-size');
  });
});

describe('telepath: wagtail.blocks.StructBlock in stream block', () => {
  let boundBlock;

  beforeEach(() => {
    // Setup test blocks - StreamBlock[StructBlock[StreamBlock[FieldBlock], FieldBlock]]
    const innerStreamDef = new StreamBlockDefinition(
      'inner_stream',
      [
        [
          '',
          [
            new FieldBlockDefinition(
              'test_block_a',
              new DummyWidgetDefinition('Block A Widget'),
              {
                label: 'Test Block A',
                required: false,
                icon: 'pilcrow',
                classname:
                  'w-field w-field--char_field w-field--admin_auto_height_text_input',
              },
            ),
          ],
        ],
      ],
      {},
      {
        label: 'Inner Stream',
        required: false,
        icon: 'placeholder',
        classname: null,
        helpText: '',
        helpIcon: '',
        maxNum: null,
        minNum: null,
        blockCounts: {},
        strings: {
          MOVE_UP: 'Move up',
          MOVE_DOWN: 'Move down',
          DELETE: 'Delete',
          DUPLICATE: 'Duplicate',
          ADD: 'Add',
        },
      },
    );

    const structBlockDef = new StructBlockDefinition(
      'struct_block',
      [
        innerStreamDef,
        new FieldBlockDefinition(
          'test_block_b',
          new DummyWidgetDefinition('Block A Widget'),
          {
            label: 'Test Block B',
            required: false,
            icon: 'pilcrow',
            classname: '',
          },
        ),
      ],
      {
        label: 'Heading block',
        required: false,
        icon: 'title',
        classname: 'struct-block',
        helpText: 'use <strong>lots</strong> of these',
        helpIcon: '<div class="icon-help">?</div>',
      },
    );

    const blockDef = new StreamBlockDefinition(
      '',
      [['', [structBlockDef]]],
      {},
      {
        label: '',
        required: true,
        icon: 'placeholder',
        classname: null,
        helpText: 'use <strong>plenty</strong> of these',
        helpIcon: '<div class="icon-help">?</div>',
        maxNum: null,
        minNum: null,
        blockCounts: {},
        strings: {
          MOVE_UP: 'Move up',
          MOVE_DOWN: 'Move down',
          DELETE: 'Delete',
          DUPLICATE: 'Duplicate',
          ADD: 'Add',
        },
      },
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render($('#placeholder'), 'the-prefix', [
      {
        type: 'struct_block',
        id: 'struct-block-1',
        value: {
          inner_stream: [
            { type: 'test_block_a', id: 'very-nested-1', value: 'foobar' },
          ],
          test_block_b: 'hello, world',
        },
      },
    ]);
  });

  test('ids are not duplicated when duplicating struct blocks', () => {
    boundBlock.children[0].duplicate();

    const duplicatedStreamChild = boundBlock.children[1];
    const originalStreamChild = boundBlock.children[0];

    expect(duplicatedStreamChild).not.toHaveSameBlockIdAs(originalStreamChild);

    const duplicatedStreamBlockInStruct =
      duplicatedStreamChild.block.childBlocks.inner_stream;
    const originalStreamBlockInStruct =
      originalStreamChild.block.childBlocks.inner_stream;

    expect(duplicatedStreamBlockInStruct.children[0]).not.toHaveSameBlockIdAs(
      originalStreamBlockInStruct.children[0],
    );
  });
});
