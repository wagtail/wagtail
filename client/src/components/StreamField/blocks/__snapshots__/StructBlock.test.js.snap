// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`telepath: wagtail.blocks.StructBlock it renders correctly 1`] = `
"<div class=\\"struct-block\\">
        
          <div class=\\"c-sf-help\\">
            <div class=\\"help\\">
              use <strong>lots</strong> of these
            </div>
          </div>
        <div data-contentpath=\\"heading_text\\">
          <label class=\\"w-field__label\\" for=\\"the-prefix-heading_text\\">Heading text<span class=\\"w-required-mark\\">*</span></label>
            <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-heading_text-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-heading_text\\" id=\\"the-prefix-heading_text\\">Heading widget</p>
          </div>
          <div id=\\"the-prefix-heading_text-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
          </div><div data-contentpath=\\"size\\">
          <label class=\\"w-field__label\\" for=\\"the-prefix-size\\">Size</label>
            <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--choice_field w-field--select\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-size-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-size\\" id=\\"the-prefix-size\\">Size widget</p>
          </div>
          <div id=\\"the-prefix-size-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
          </div></div>"
`;

exports[`telepath: wagtail.blocks.StructBlock setError passes error messages to children 1`] = `
"<div class=\\"struct-block\\">
        
          <div class=\\"c-sf-help\\">
            <div class=\\"help\\">
              use <strong>lots</strong> of these
            </div>
          </div>
        <div data-contentpath=\\"heading_text\\">
          <label class=\\"w-field__label\\" for=\\"the-prefix-heading_text\\">Heading text<span class=\\"w-required-mark\\">*</span></label>
            <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-heading_text-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-heading_text\\" id=\\"the-prefix-heading_text\\">Heading widget</p>
          </div>
          <div id=\\"the-prefix-heading_text-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
          </div><div data-contentpath=\\"size\\">
          <label class=\\"w-field__label\\" for=\\"the-prefix-size\\">Size</label>
            <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--choice_field w-field--select w-field--error\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-size-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\"><use href=\\"#icon-warning\\"></use></svg>
          <p class=\\"error-message\\"><span>This is too big</span></p></div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-size\\" id=\\"the-prefix-size\\">Size widget</p>
          </div>
          <div id=\\"the-prefix-size-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
          </div></div>"
`;

exports[`telepath: wagtail.blocks.StructBlock setError shows non-block errors 1`] = `
"<div class=\\"struct-block\\"><p class=\\"help-block help-critical\\">This is just generally wrong</p>
        
          <div class=\\"c-sf-help\\">
            <div class=\\"help\\">
              use <strong>lots</strong> of these
            </div>
          </div>
        <div data-contentpath=\\"heading_text\\">
          <label class=\\"w-field__label\\" for=\\"the-prefix-heading_text\\">Heading text<span class=\\"w-required-mark\\">*</span></label>
            <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-heading_text-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-heading_text\\" id=\\"the-prefix-heading_text\\">Heading widget</p>
          </div>
          <div id=\\"the-prefix-heading_text-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
          </div><div data-contentpath=\\"size\\">
          <label class=\\"w-field__label\\" for=\\"the-prefix-size\\">Size</label>
            <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--choice_field w-field--select\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-size-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-size\\" id=\\"the-prefix-size\\">Size widget</p>
          </div>
          <div id=\\"the-prefix-size-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
          </div></div>"
`;

exports[`telepath: wagtail.blocks.StructBlock with formTemplate it renders correctly 1`] = `
"<div class=\\"custom-form-template\\">
        <p>here comes the first field:</p>
        <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--char_field w-field--text_input\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-heading_text-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-heading_text\\" id=\\"the-prefix-heading_text\\">Heading widget</p>
          </div>
          <div id=\\"the-prefix-heading_text-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
        <p>and here is the second:</p>
        <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <div class=\\"w-field w-field--choice_field w-field--select\\" data-field=\\"\\">
          <div class=\\"w-field__errors\\" id=\\"the-prefix-size-errors\\" data-field-errors=\\"\\">
            <svg class=\\"icon icon-warning w-field__errors-icon\\" aria-hidden=\\"true\\" hidden=\\"\\"><use href=\\"#icon-warning\\"></use></svg>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <p name=\\"the-prefix-size\\" id=\\"the-prefix-size\\">Size widget</p>
          </div>
          <div id=\\"the-prefix-size-helptext\\" data-field-help=\\"\\"></div>
        </div>
      </div>
      </div>"
`;
